### PR TITLE
Convert wan/3gpp-scan test as a automated test (New)

### PR DESCRIPTION
**Pull request by stanley31huang**
## Description
Convert wan/3gpp-scan test as a automated test (New)

## Resolved issues
[OEMQA-4494](https://warthogs.atlassian.net/browse/OEMQA-4494)

## Documentation
N/A

## Tests
Please refer following submission for the test results (w/ side loaded provider)
https://certification.canonical.com/hardware/202302-31253/submission/379312/ (Updated on 2024/07/08)

[OEMQA-4494]: https://warthogs.atlassian.net/browse/OEMQA-4494?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: providers/base/units/wwan/test-plan.pxu [hunk 1]
```diff
@@ -16,6 +16,7 @@ _description: Automated wwan tests for Snappy Ubuntu Core devices
 include:
     # Note these tests require snap calling snap support
     wwan/detect
+    wwan/3gpp-scan-.*-auto
```
Comment (pieqq) [inline providers/base/units/wwan/test-plan.pxu:19]: Use the `template-id` instead of regex.
```suggestion
    wwan/3gpp-scan-manufacturer-model-hw_id-auto
```

File: providers/base/units/wwan/test-plan.pxu [hunk 2]
```diff
@@ -36,6 +37,8 @@ _name: Automated wwan tests (after suspend)
 _description: Automated wwan tests for Snappy Ubuntu Core devices
 include:
     after-suspend-wwan/detect
+    after-suspend-wwan/check-sim-present-.*-auto
+    after-suspend-wwan/3gpp-scan-.*-auto
```
Comment (pieqq) [inline providers/base/units/wwan/test-plan.pxu:41]: Same here:

```suggestion
    after-suspend-wwan/check-sim-present-manufacturer-model-hw_id-auto
    after-suspend-wwan/3gpp-scan-manufacturer-model-hw_id-auto
```

File: providers/base/bin/wwan_tests.py [hunk 3]
```diff
@@ -362,6 +362,45 @@ def invoked(self):
...
+            _wwan_radio_on()
+            cmd = [
+                "mmcli",
+                "-m",
+                str(mm_id),
+                "--3gpp-scan",
+                "--timeout",
+                str(args.timeout),
+            ]
+            ret_code = subprocess.check_call(cmd)
```
Comment (pieqq) [inline providers/base/bin/wwan_tests.py:397]: Using `check_call`, you will only get back the return code of the `mmcli` command. It's fine if there is no useful feedback we could get from the output, but otherwise it might be interesting to capture the output in case the command fails. (I forgot what output `mmcli --3gpp-scan` provides, so I'm not sure if my comment makes sense :))

File: providers/base/tests/test_wwan_tests.py [hunk 4]
```diff
@@ -109,3 +111,98 @@ def test_invoked_with_mmdbus(self, mock_mmdbus):
...
+            obj_3gppscan.invoked()
+
+        self.assertEqual(context.exception.code, 1)
+        self.assertTrue(mock_mmcli.called)
+        self.assertTrue(mmcli_instance.equipment_id_to_mm_id.called)
+        self.assertTrue(mock_r_on.called)
+        self.assertTrue(mock_r_off.called)
+        mock_subprocess.assert_called_with(
+            ["mmcli", "-m", "0", "--3gpp-scan", "--timeout", "200"]
+        )
```
Comment (pieqq) [inline providers/base/tests/test_wwan_tests.py:180]: Given that you also test when a `CalledProcessError` is raised, I don't think this test is required, as it covers the same parts of your code.

(even if you remove it, the coverage for `wwan_tests.py` will still be 100%)

**Review by pieqq — CHANGES_REQUESTED** [review]
A few comments below, and a change request to use `template-id`s instead of regex in the test plans.

File: providers/base/tests/test_wwan_tests.py (hunk 4 above)
Comment (stanley31huang) [reply providers/base/tests/test_wwan_tests.py:180]: no, the coverage will down to 89%. So that's why I added this unit test.

File: providers/base/bin/wwan_tests.py (hunk 3 above)
Comment (stanley31huang) [reply providers/base/bin/wwan_tests.py:397]: Here is some errors we might seen during the scanning network test, so we probably could check is `error` shown in the log.
```
# the modem is not available
admin@C0310A2043-00843:~$ sudo mmcli -m 0 --3gpp-scan --timeout 200; echo $?
error: couldn't find modem
1

# the modem is not enabled
admin@C0310A2043-00843:~$ sudo mmcli -m 1 --3gpp-scan --timeout 200; echo $?
error: modem not enabled yet
1

# Unable to scan available network within a period of time
admin@C0310A2043-00843:~$ sudo mmcli -m 1 --3gpp-scan --timeout 30; echo $?
error: couldn't scan networks in the modem: 'Timeout was reached'
1
```

File: providers/base/tests/test_wwan_tests.py (hunk 4 above)
Comment (stanley31huang) [reply providers/base/tests/test_wwan_tests.py:180]: here are some error symptom while scanning 3GPP network. So we could simply check if the `error` message exists or not.
```
# the modem is not enabled
admin@C0310A2043-00843:~$ sudo mmcli -m 1 --3gpp-scan --timeout 200; echo $?
error: modem not enabled yet
1

# Failed to scan available network within 100s
admin@C0310A2043-00843:~$ sudo mmcli -m 0 --3gpp-scan --timeout 100; echo $?
error: couldn't scan networks in the modem: 'Timeout was reached'
1
```

File: providers/base/bin/wwan_tests.py [hunk 5]
```diff
@@ -383,21 +412,29 @@ def invoked(self):
...
+                ]
+                print_head("Scanning 3GPP available network")
+                print("+ {}".format(" ".join(cmd)))
+                ret_sp = subprocess.run(
+                    cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+                )
+                print(ret_sp.stdout.decode("utf-8"))
+                if (
+                    ret_sp.returncode == 0
+                    and b"error" not in ret_sp.stdout.lower()
```
Comment (pieqq) [inline providers/base/bin/wwan_tests.py:432]: Why is this needed? Are there cases where the return code from the 3gpp scanning is `0` but there is still an error?

Comment (stanley31huang) [reply providers/base/bin/wwan_tests.py:432]: just in case that the `3gpp-scan` will always return `0` no matter the connection is detected or not.

BTW, here is the output while the system can't scan any available connection. 
```
admin@C0310A2043-00843:~$ sudo mmcli -m 0 --3gpp-scan --timeout 200; echo $?
error: couldn't scan networks in the modem: 'Timeout was reached'
1
```

File: providers/base/bin/wwan_tests.py [hunk 6]
```diff
@@ -362,6 +372,72 @@ def invoked(self):
...
+                ):
+                    ret_code = 0
```
Comment (stanley31huang) [inline providers/base/bin/wwan_tests.py:434]:
```suggestion
                ret_code = ret_sp.returncode
```

File: providers/base/bin/wwan_tests.py (hunk 5 above)
Comment (stanley31huang) [reply providers/base/bin/wwan_tests.py]: Thanks @pieqq  for the suggestion, I have sent a fixed.

**Review by pieqq — APPROVED** [review]
Looks good to me, thanks for applying all the changes we discussed here and on Mattermost!
